### PR TITLE
HD key support (5.0.4)

### PR DESCRIPTION
**Pull request by anonymoussprocket**
Added support for hierarchically deterministic keys for the ED25519 curve. Also in this update:

- SECP256K1 signature verification.
- documentation update.

### CODE REVIEW

File: src/KeyStoreUtils.ts [hunk 1]
```diff
@@ -57,8 +59,17 @@ export namespace KeyStoreUtils {
             if (!bip39.validateMnemonic(mnemonic)) { throw new Error('The given mnemonic could not be validated.'); }
         }
 
-        const seed = (await bip39.mnemonicToSeed(mnemonic, password)).slice(0, 32);
-        const keys = await generateKeys(seed);
+        let keys: { secretKey: Buffer, publicKey: Buffer };
+        const seed = await bip39.mnemonicToSeed(mnemonic, password);
+        if (derivationPath !== undefined) {
+            const sk = (Ed25519.derivePath(derivationPath, seed.toString("hex"))).key;
```
Comment (itamarreif) [inline src/KeyStoreUtils.ts]: why do you need both sk and p? seems sk isnt used

File: src/KeyStoreUtils.ts [hunk 2]
```diff
@@ -134,9 +145,35 @@ export namespace KeyStoreUtils {
...
+
+        const sig = TezosMessageUtils.writeSignatureWithHint(signature, sigPrefix);
+        const pk = TezosMessageUtils.writeKeyWithHint(publicKey, keyPrefix);
+
+        if (curve === SignerCurve.ED25519) {
+            return await CryptoUtils.checkSignature(sig, bytes, pk);
+        }
+
+        if (curve === SignerCurve.SECP256K1) {
+            return secp256k1.ecdsaVerify(sig, bytes, pk);
```
Comment (itamarreif) [inline src/KeyStoreUtils.ts:181]: maybe also wrap this in CryptoUtils

Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts:181]: CryptoUtils is a wrapper for libsodium mostly, which doesn't support this curve.

File: src/KeyStoreUtils.ts (hunk 1 above)
Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts]: Thank you

File: src/KeyStoreUtils.ts (hunk 2 above)
Comment (itamarreif) [reply src/KeyStoreUtils.ts:181]: oh, thought CryptoUtils was vendoring the crypto libraries

Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts:181]: We mostly just work with ed25519 for tz1 accounts.

**Review by itamarreif — DISMISSED** [review]

File: src/KeyStoreUtils.ts [hunk 3]
```diff
@@ -57,8 +59,16 @@ export namespace KeyStoreUtils {
             if (!bip39.validateMnemonic(mnemonic)) { throw new Error('The given mnemonic could not be validated.'); }
         }
 
-        const seed = (await bip39.mnemonicToSeed(mnemonic, password)).slice(0, 32);
-        const keys = await generateKeys(seed);
+        let keys: { secretKey: Buffer, publicKey: Buffer };
+        const seed = await bip39.mnemonicToSeed(mnemonic, password);
```
Comment (vishakh) [inline src/KeyStoreUtils.ts:63]: How come we don't slice anymore?

File: src/KeyStoreUtils.ts [hunk 4]
```diff
@@ -57,8 +59,16 @@ export namespace KeyStoreUtils {
...
+        if (derivationPath !== undefined && derivationPath.length > 0) {
+            const p = Ed25519.derivePath(derivationPath, seed.toString("hex"));
```
Comment (vishakh) [inline src/KeyStoreUtils.ts]: Could we make this variable name more descriptive?

File: src/KeyStoreUtils.ts [hunk 5]
```diff
@@ -57,8 +59,16 @@ export namespace KeyStoreUtils {
...
+
+            keys = await recoverKeys(Buffer.concat([p.key, p.chainCode]));
```
Comment (vishakh) [inline src/KeyStoreUtils.ts]: For maximum clarity, can we make `Buffer.concat([p.key, p.chainCode]` a named const?

File: src/KeyStoreUtils.ts [hunk 6]
```diff
@@ -134,9 +144,35 @@ export namespace KeyStoreUtils {
             messageBytes = Buffer.from(message, 'utf8');
         }
 
-        const sig = TezosMessageUtils.writeSignatureWithHint(signature, 'edsig');
-        const pk = TezosMessageUtils.writeKeyWithHint(publicKey, 'edpk');
+        return checkSignature(signature, messageBytes, publicKey);
+    }
+
+    export async function checkSignature(signature: string, bytes: Buffer, publicKey: string): Promise<boolean> {
```
Comment (vishakh) [inline src/KeyStoreUtils.ts:158]: Let's document this function since it will be used by third parties.

File: src/KeyStoreUtils.ts (hunk 3 above)
Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts:63]: we still do, line 69, but that's how it works for non-derivation-path keys

File: src/KeyStoreUtils.ts (hunk 4 above)
Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts]: P, for Path!

Comment (anonymoussprocket) [reply src/KeyStoreUtils.ts]: new commit coming...

**Review by itamarreif — APPROVED** [review]

**Review by vishakh — APPROVED** [review]

**Pull request merged**